### DISCUSSION TITLE
[@types/meteor] Meteor.User profile makes overriding impossible

### DISCUSSION THREAD

**Discussion by ebroder** [DefinitelyTyped/DefinitelyTyped#58366] [Issues with a @types package] [answered]
The current declaration of the `profile` property on `Meteor.User` (it's currently declared as [`profile?: any`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/21c9f99f8e0f62f0fef7194e52e8782c8c737e0a/types/meteor/meteor.d.ts#L49)) makes it somewhat difficult to use. In my app, I'd like to declare it to have a somewhat more specific type.

I tried adding a type declaration to my app with something like:

```ts
declare module 'meteor/meteor' {
  module Meteor {
    interface User {
      profile?: {
        myField: string;
        etc: number;
      }
    }
  }
}
```

but that throws a compilation error (`Subsequent property declarations must have the same type.  Property 'profile' must be of type 'any', but here has type '{ …`). Sure, makes sense.

Is there a different declaration that would allow downstream developers to set their own type for `profile`? The best idea I have is to make `User` generic, so something like:

```ts
interface User<T> {
  profile?: T
}
```

and then redeclare everything that interacts with `User` to also take a generic parameter. That seems like a miserable compatibility break, and not obviously possible (I'm not sure how to declare `Meteor.users` under that paradigm.

Is there something reasonable we can change the declaration to that would allow using it downstream?

(As an aside: I'm aware that the default Meteor configuration makes `profile` directly-writeable by clients, and thus `any` is not an _inaccurate_ declaration. However, Meteor specifically [recommends](https://docs.meteor.com/api/accounts.html#:~:text=To%20forbid%20users%20from%20making%20any%20modifications%20to%20their%20user%20document%3A) a configuration that blocks that default, which is what I'm using in my application.)

[a comment by an automated account is not shown]

**Reply by edemaine** [reply]
It seems like a weak type like `profile?: object` would make sense. This would accurately reflect that, if `profile` exists, it's an object, but might not have any specific keys you were expecting (unlike `any`).

Can you confirm that this would be overridable by your specific object type? (I think so but haven't tested.)

**Reply by ebroder** [reply]
Hmm, doesn't seem like that works - I still get an error for redeclaring the type. Here's a [playground example](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKoGdoFEC2dgA2yA3gFDIXJwAm1UE66AXMumFKAOYDc5lAbtGAxgEaiwBGAeykEIcELwC+pUJFiIUGaCT4UA+sHGt2XXpWQBXTFBBwcEAPws2HEJ2QAfKyGoQRIGLmlBB4hOjOaDa4+AQA2gC6Xj5+AUF6yAj0cJDUAIJgkQAiOSjelr7+oOkWAA5QUiJykVISAFYQCGDBFDb8wEgRLAoAnsqkpAD0k8hYAB6QvlHQE2rQ8EjLULp1DU1OLGQWFnYOka5mGUrjQA)

**Reply by edemaine** [reply]
Good point. Instead perhaps we should provide an empty interface `UserProfile` that can be overridden, as in [this Playground example](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKoGdoFEC2dgA2yA3gFDIXJwAm1UE66AXMumFKAOYDc5lAbtGAxgEaiwBGAeykEIcELwC+pUJFiIUGaAAUoUkXJKkVa6PCRpMUY5WQB9YONbsuvOwFdrIODggB+FjYOEE5kAB9kDxBqCBEQMXdKCDxCdECrbFSCAG0AXQiomLjQRL4KBHo4SGoAQTAMgBFqlEjo2PiyuwAHfUMAlm0oPQNCCCSKa35gJHSWBQBPZVJSAHpV5CwAD0gYzKgVsw1LIZH+20ofPwzgtxNSIA).

**Comment by ebroder** [comment, marked as answer]
@edemaine recommended creating an empty `UserProfile` interface that can be overridden by app developers, and I think that's a good answer, so I'll push a pull request with that shortly. Thanks for the suggestion!
